docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1844,16 +1855,94 @@ entry value. When providing serializable enum values through `P4Data`, one can
 either use the enum entry's name (`enum` human-readable string field) or its
 assigned value (`enum_value` bytestring field).
 
+### User-defined types { #sec-user-defined-types}
+
+P4~16~ enables programmers to introduce new types [@P4NewTypes]. While similar
+to `typedef`, this mechanism introduces in fact a new type, which is not
+strictly synonym with the original type. It is important to preserve this
+distinction in the P4Info message, in particular for the purposes of
+[translation](#sec-psa-metadata-translation). When introducing a new type, the
+declaration can be annotated with `@p4runtime_translation` to indicate that the
+type exposed to the P4Runtime client is different from the original P4 type. One
+important use-case if for [port numbers](#sec-translation-of-port-numbers),
+whose underlying dataplane representation may vary on different targets, but for
+which it may be convenient to present a unified representation and numbering
+scheme to the control-plane. **The `@p4runtime_translation` annotation can only
+be used if the underlying P4 built-in type is a fixed-width unsigned bitstring
+type (`bit<W>`)** and the type exposed to the control-plane will also be a
+fixed-width unsigned bitstring, with a potentially different bitwidth. It takes
+two parameters: a *URI* (Uniform Resource Identifier) which uniquely identifies
+the translation being performed on entities of the new type to the P4Runtime
+server and the *bitwidth* of the bitstring type exposed to the control-plane. It
+is recommended that the URI includes at least the P4 architecture name and the
+type name.
+
+User-defined types are specified using the `P4NewTypeSpec` message, which has
+the following fields:
+
+* `original_type`, which specifies the underlying type used in the P4 `type`
+  declaration, using the generic `P4DataTypeSpec` message. The `original_type`
+  can correspond to a P4~16~ built-in type or to another user-defined type.
```

Comment:
As I think Antonin commented elsewhere, at least for the current 7 PSA types that we want to allow to have bit widths that can vary across targets, including `original_type` for those would conflict with the goal of making the P4Info file the same for all targets.

The set of types with that property we expect to change very slowly, if at all (on the time scale of releases of new versions of the PSA spec).

One possibility: For such types define a special value for the `original_type` field that is easily distinguishable from all other type definitions, which means "target-dependent", with no further details.